### DISCUSSION TITLE
No longer downloading

### DISCUSSION THREAD

**Discussion by TekyTech** [vicwomg/pikaraoke#350] [General]
I had a fully functioning installation on LinuxLite 7.0. I suggested a feature that was reported as implemented and I wanted to see how it works now. 
I wasn’t sure about the best way to update to the most current build. So I ran all Linux updates, ran git to download the latest build and I ran the setup again same as if it was first installation, thinking that it would override  whatever has changed. 
I didn’t see any change about the feature that I suggested which was marked as complete. 

_**PROBLEM**_:
It no longer downloads anything. 
It finds songs ok when doing search
It plays anything from the library of everything that was previously downloaded. 
It no longer downloads anything new 
There are no error, messages, but the download just doesn’t happen  
I tried to go to the web interface where it says update dl. Nothing changes either. 

Any ideas?

Thank you.

**Comment by vicwomg** [comment]
The proper way to update to the latest main branch is to simply `git pull` from your pikaraoke dir

- What version of youtube-dl does the info screen report you're using? 
- Run with debug logging `./pikaraoke.sh -l10`  and see if it says anything useful during a download attempt

**Reply by TekyTech** [reply]
My apologies. I wiped out the system and completely reinstalled everything from scratch before i saw this. So I cannot troubleshoot it anymore. The new installation works beautifully now. Everything downloads and the new way to show a list of songs is really great!

**Comment by TekyTech** [comment]
Hello. I had three laptops for relatives to use karaoke at three households. A similar thing happened to another laptop. I had everything fully functional. It had Lubuntu, relatively recent, not too old. They ran Lubuntu updates. Nobody tried to reinstall pikaraoke. It just stopped downloading. Quietly, no error messages. Searches just fine. After click on Download just nothing happens anymore. 

That one was installed without git so I cannot use git update. I simply downloaded the installation package from GitHub and ran setup.sh from there. It actually worked that way. 

By the way, I tried to update YouTube dl.  No change. 

QUESTION, if I may:
In this situation, what would you suggest would be the best way to update/upgrade installation? Or how should I uninstall it to cleanly install pikaraoke again? I would prefer not to wipe out Lubuntu installation. That laptop was a little bit challenging and installing Ubuntu or LinuxLite  didn’t work so I had to settle with the only distro that worked on it - Lubuntu. 

Anyway, what’s the best path forward? Any suggestions?

I’m ok to uninstall and reinstall pikaraoke from scratch, using git properly. Then no need to worry about updating pikaraoke. Is this the best way? If yes, should I uninstall current installation? How is the best to do it?

If nothing helps of course I can wipe out Lubuntu. Just prefer not to. 

Any suggestions are much appreciated.

**Reply by vicwomg** [reply]
You should first trouble shoot the actual download problem. Downloading issues usually have to do with the yt-dl. What version does it say is installed when you have a look at the info page? Run pikaraoke with the `-l10` option to turn on debug logging and observe the log output when a download is attempted.

Installing via `git clone` (see the readme) and updating via `git pull` is preferred to pick up the most recent updates. You should not have to reinstall the entire OS. If you use the zipped release artifacts you'll simply have to wait until the next official release. You can even add the git pull command to your startup script if you really want to be hands off about the whole thing.

**Reply by vicwomg** [reply]
Note that the latest version of yt-dlp is pretty fresh. Came out 4 days ago. [2024.07.25](https://github.com/yt-dlp/yt-dlp/releases/tag/2024.07.25)

If the installed version is older than that it probably needs an update. If the update function doesn't work you need to troubleshoot why. Again, debug logging helps.

**Reply by TekyTech** [reply]
Thank you so much for trying to help and replying so fast too. Much appreciated. 

Youtube-dl (yt-dlp) version: 2024.05.27

I tried to play the song from the log below (same YouTube karaoke clip URL) in a browser, it plays fine on the same computer. Nothing is “403 forbidden” in the browser. 

Here’s the error that the log shows. 
(By the way it does show errors, so if it does maybe it can be caught by error handling and an alert with an error message shown, as in a feature request I posted separately):

[2024-07-29 09:01:48] INFO: Downloading video: https://www.youtube.com/watch?v=zAbT9dr1R30
[2024-07-29 09:01:48] DEBUG: Youtube-dl command: /home/user/PiKaraoke/.venv/bin/yt-dlp -f mp4 -o /home/user/pikaraoke-songs/%(title)s---%(id)s.%(ext)s https://www.youtube.com/watch?v=zAbT9dr1R30
[youtube] Extracting URL: https://www.youtube.com/watch?v=zAbT9dr1R30
[youtube] zAbT9dr1R30: Downloading webpage
[youtube] zAbT9dr1R30: Downloading ios player API JSON
[youtube] zAbT9dr1R30: Downloading player 0e90208b
WARNING: [youtube] zAbT9dr1R30: nsig extraction failed: You may experience throttling for some formats
         n = gYE5OjjoOHne0g18 ; player = https://www.youtube.com/s/player/0e90208b/player_ias.vflset/en_US/base.js
WARNING: [youtube] zAbT9dr1R30: nsig extraction failed: You may experience throttling for some formats
         n = 7LyD_dogQEs3p8z4 ; player = https://www.youtube.com/s/player/0e90208b/player_ias.vflset/en_US/base.js
[youtube] zAbT9dr1R30: Downloading m3u8 information
[info] zAbT9dr1R30: Downloading 1 format(s): 18
ERROR: unable to download video data: HTTP Error 403: Forbidden
[2024-07-29 09:01:54] ERROR: Error code while downloading, retrying once...
[youtube] Extracting URL: https://www.youtube.com/watch?v=zAbT9dr1R30
[youtube] zAbT9dr1R30: Downloading webpage
[youtube] zAbT9dr1R30: Downloading ios player API JSON
[youtube] zAbT9dr1R30: Downloading player 0e90208b
WARNING: [youtube] zAbT9dr1R30: nsig extraction failed: You may experience throttling for some formats
         n = gNcIPYW0yegtmHp6 ; player = https://www.youtube.com/s/player/0e90208b/player_ias.vflset/en_US/base.js
WARNING: [youtube] zAbT9dr1R30: nsig extraction failed: You may experience throttling for some formats
         n = EkY4GErIncWimT4W ; player = https://www.youtube.com/s/player/0e90208b/player_ias.vflset/en_US/base.js
[youtube] zAbT9dr1R30: Downloading m3u8 information
[info] zAbT9dr1R30: Downloading 1 format(s): 18
ERROR: unable to download video data: HTTP Error 403: Forbidden
[2024-07-29 09:01:58] ERROR: Error downloading song: https://www.youtube.com/watch?v=zAbT9dr1R30
[2024-07-29 09:02:32] INFO: Clearing queue!

**Reply by TekyTech** [reply]
> Note that the latest version of yt-dlp is pretty fresh. Came out 4 days ago. [2024.07.25](https://github.com/yt-dlp/yt-dlp/releases/tag/2024.07.25)
> 
> If the installed version is older than that it probably needs an update. If the update function doesn't work you need to troubleshoot why. Again, debug logging helps.

When I click on update dl from the karaoke interface nothing changes, version stays the same. 

What’s the best way to update it manually from the terminal, please?

**Reply by TekyTech** [reply]
P.S. you are awesome :)

**Reply by vicwomg** [reply]
Your Youtube dl version  `2024.05.27` is out of date.  Current is [2024.07.25](https://github.com/yt-dlp/yt-dlp/releases/tag/2024.07.25)

Try to update it via the info screen and get the debug log output to see why it might be failing

**Reply by TekyTech** [reply]
When I click on update dl from the karaoke interface, the info page, nothing changes, version stays the same.

What’s the best way to update it manually using terminal, please?

**Reply by TekyTech** [reply]
Here’s the log of trying to update dl from the info page:

2024-07-29 09:34:57] INFO: Connect the player host to: http://172.20.1.82:5555/splash
[2024-07-29 09:35:16] INFO: Upgrading youtube-dl, current version: 2024.05.27
[2024-07-29 09:35:19] INFO: Current version: stable@2024.05.27 from yt-dlp/yt-dlp
Latest version: stable@2024.07.25 from yt-dlp/yt-dlp
ERROR: You installed yt-dlp with pip or using the wheel from PyPi; Use that to update

[2024-07-29 09:35:19] INFO: Attempting youtube-dl upgrade via pip3...
ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: '/home/user/PiKaraoke/.venv/bin/yt-dlp'
Check the permissions.                                                                                                                         
                                                                                                                                               
Exception in thread Thread-102 (update_youtube_dl):                                                                                            
Traceback (most recent call last):
  File "/usr/lib/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.12/threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user/PiKaraoke/app.py", line 577, in update_youtube_dl
    k.upgrade_youtubedl()
  File "/home/user/PiKaraoke/karaoke.py", line 226, in upgrade_youtubedl
    output = check_output(
             ^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pip3', 'install', '--upgrade', 'yt-dlp']' returned non-zero exit status 1.
[2024-07-29 09:35:34] INFO: Clearing queue!

**Reply by vicwomg** [reply]
The error is right there.

```
ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: '/home/user/PiKaraoke/.venv/bin/yt-dlp'
Check the permissions.
```

The user doesn't have permission to update. It's a linux permissions issue. If you installed pikaraoke as root and are running it as an under-privileged user. Sounds like an issue for your sysadmin (probably you)

**Comment by TekyTech** [comment]
By the way I tried the following from terminal:
pip install --upgrade yt-dlp
and:
yt-dlp -U 

First says that the environment is externally managed. 

Second says yt-dlp not found. 

Let me try with snap…

**Reply by vicwomg** [reply]
This wont work. Youtube dl is installed and isolated within pikaraoke's virtual environment, not globally. The issue is that of OS permissions.

**Reply by TekyTech** [reply]
Yes thank you. It seems I found it out too because I just installed the latest version of yt-dlt using snap but of course pikaraoke info page still shows older version. 

Also I tried to start pikaraoke.sh with sudo and it crashes. 

Any better ideas other than installing it again, maybe with git this time.

**Reply by vicwomg** [reply]
I'm not a linux expert, and I don't really know how you installed everything and I don't know your user heirarchy. You probably need to do some of your own legwork in figuring out why your permissions aren't matching. It sounds to me like whatever user you're logged in as does not have proper permissions. 

`whoami` will tell you your current user
`ls -la /home/user/PiKaraoke/.venv` will list the owner of the .venv directory and the r/w configuration. 

If those don't match, your current user doesn't have the permission to update. Some combo of `chown` or `chmod` might fix it, but again, not a linux expert. You're on your own there.

Otherwise, yes, delete the pikaraoke directory and install as recommended in the README via git.

**Comment by TekyTech** [comment]
I would like to thank you for your suggestions. 
Family is happy!

I installed new instance properly, using git. It went smooth except what I knew from before - I needed to uninstall chromium right after installation, otherwise it would be stuck on a blank while screen. Everything is working beautifully now again. 

Oh and I added git pull to the startup before starting pikaraoke. 

So git pull supposed to also update yt-dl next time? Or only through pikaraoke info page?

I guess nothing from the old installation running on the background?

I deleted old folder. Nothing to uninstall I guess and it’s clean enough and won’t leave any unwanted traces?

Thank you again from me and our extended family. You rock!
